### PR TITLE
Make Transpose's perm argument optional (by default, reverse dims)

### PR DESCRIPTION
**Pull request by jantonguirao**
Signed-off-by: Joaquin Anton <janton@nvidia.com>

## Category:

**New feature**

## Description:

- Makes transpose's perm argument optional, introducing a default permutation
- Default permutation is to reverse the dims

### Affected modules and functionalities:

Transpose

### Tests:

test_transpose.py : test_transpose_default
- [ ] Existing tests apply
- [x] New tests added
  - [x] Python tests
  - [ ] GTests
  - [ ] Benchmark
  - [ ] Other
- [ ] N/A

### Documentation
- [ ] Existing documentation applies
- [x] Documentation updated
  - [x] Docstring
  - [ ] Doxygen
  - [ ] RST
  - [ ] Jupyter
  - [ ] Other
- [ ] N/A

#### Requirements
- [x] Implements new requirements
- [ ] Affects existing requirements
- [ ] N/A

**REQ IDs**: TRPOSE.06

**JIRA TASK**: N/A

### CODE REVIEW

File: dali/operators/generic/transpose/transpose.h [hunk 1]
```diff
@@ -74,18 +79,27 @@ class Transpose : public Operator<Backend> {
...
-    int dim = input_shape.sample_dim();
-    int pdim = perm_.size();
-    DALI_ENFORCE(dim == pdim, make_string("Input has different dimensionality (", dim,
-        ") than the length of the permutation (", pdim, ")"));
+    int ndim = input_shape.sample_dim();
+
+    if (default_perm_ && static_cast<int>(perm_.size()) != ndim) {
+      perm_.resize(ndim);
+      std::iota(perm_.begin(), perm_.end(), 0);
+      std::reverse(perm_.begin(), perm_.end());
```
Comment (mzient) [inline dali/operators/generic/transpose/transpose.h:88]: Now that you have to reverse, a simple loop would be clearer.

File: dali/operators/generic/transpose/transpose.h [hunk 2]
```diff
@@ -74,18 +79,27 @@ class Transpose : public Operator<Backend> {
...
     output_desc[0].type = input.type();
-
+    output_desc[0].shape = input_shape;
+
+    if (!perm_.empty()) {
+      int dim = input_shape.sample_dim();
+      int pdim = perm_.size();
+      DALI_ENFORCE(dim == pdim, make_string("Input has different dimensionality (", dim,
+          ") than the length of the permutation (", pdim, ")"));
+      permute_dims(output_desc[0].shape, input_shape, perm_);
```
Comment (mzient) [inline dali/operators/generic/transpose/transpose.h:101]: What happened before? Some error later in the kernel?

File: dali/operators/generic/transpose/transpose_gpu.cc [hunk 3]
```diff
@@ -40,10 +40,10 @@ class TransposeGPU : public Transpose<GPUBackend> {
   bool SetupImpl(vector<OutputDesc> &descs, const Workspace &ws) override {
     Transpose<GPUBackend>::SetupImpl(descs, ws);
     const auto &input = ws.Input<GPUBackend>(0);
-
+    const auto& shape = input.shape();
```
Comment (mzient) [inline dali/operators/generic/transpose/transpose_gpu.cc:43]: Total nitpick ;)
```suggestion
    const auto &shape = input.shape();
```

**Review by mzient — APPROVED** [review]

File: dali/operators/generic/transpose/transpose.h (hunk 2 above)
Comment (jantonguirao) [reply dali/operators/generic/transpose/transpose.h:101]: No, this was just moved. I moved to after I potentially regenerate the perm values.

**Review by banasraf — APPROVED** [review]

**Comment by jantonguirao** [comment]
!build

**Pull request merged**